agent/exec/containerd/controller.go

```diff
@@ -0,0 +1,366 @@
+		// same point after cancellation.
+		var pctx context.Context
+
+		r.pulled = make(chan struct{})
+		pctx, r.cancelPull = context.WithCancel(context.Background()) // TODO(stevvooe): Bind a context to the entire controller.
+
+		go func() {
+			defer close(r.pulled)
+			r.pullErr = r.adapter.pullImage(pctx)
+		}()
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.pulled:
+		if r.pullErr != nil {
+			// NOTE(stevvooe): We always try to pull the image to make sure we have
+			// the most up to date version. This will return an error, but we only
+			// log it. If the image truly doesn't exist, the create below will
+			// error out.
+			//
+			// This gives us some nice behavior where we use up to date versions of
+			// mutable tags, but will still run if the old image is available but a
+			// registry is down.
+			//
+			// If you don't want this behavior, lock down your image to an
+			// immutable tag or digest.
+			log.G(ctx).WithError(r.pullErr).Error("pulling image failed")
+		}
+	}
+
+	if err := r.adapter.create(ctx); err != nil {
+		if isContainerCreateNameConflict(err) {
+			if _, err := r.adapter.inspect(ctx); err != nil {
+				return err
+			}
+
+			// container is already created. success!
+			return exec.ErrTaskPrepared
+		}
+
+		return errors.Wrap(err, "create container failed")
+	}
+
+	return nil
+}
+
+// Start the container. An error will be returned if the container is already started.
+func (r *controller) Start(ctx context.Context) error {
+	ctx = log.WithModule(ctx, "containerd")
+
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	ctnr, err := r.adapter.inspect(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Detect whether the container has *ever* been started. If so, we don't
+	// issue the start.
+	//
+	// TODO(stevvooe): This is very racy. While reading inspect, another could
+	// start the process and we could end up starting it twice.
+	if ctnr.Status != task.StatusCreated {
+		return exec.ErrTaskStarted
+	}
+
+	if err := r.adapter.start(ctx); err != nil {
+		return errors.Wrap(err, "starting container failed")
+	}
+
+	// TODO(ijc): Wait for HealtCheck to report OK.
+
+	return nil
+}
+
+// Wait on the container to exit.
+func (r *controller) Wait(ctx context.Context) error {
+	ctx = log.WithModule(ctx, "containerd")
+
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	// check the initial state and report that.
+	ctnr, err := r.adapter.inspect(ctx)
+	if err != nil {
+		return errors.Wrap(err, "inspecting container failed")
+	}
+
+	shutdownWithExitStatus := func(reason string) error {
+		exitStatus, err := r.adapter.shutdown(ctx)
+		if err != nil {
+			return err
+		}
+		log.G(ctx).Errorf("EXIT STATUS %v", exitStatus)
+		return makeExitError(exitStatus, reason)
+	}
+	switch ctnr.Status {
+	case task.StatusStopped:
+		return shutdownWithExitStatus("")
+	}
+
+	eventq, closed, err := r.adapter.events(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.G(ctx).Infof("Wait")
```

Comment:
FTR it was more informative, since `ctx` carries various things including the task id and includes them in the log messages. I've removed it anyway though.